File: src/tools/mcpTools.ts
import { DynamicStructuredTool } from "@langchain/core/tools";
import { z } from "zod";
import {
    addPaginationInfo,
    fetchAllRecipeNames,
    fetchRecipesByDietWithPagination,
    handleDietSearch,
    handleIngredientSearch,
    handleRecipeSearchFallback,
    type RecipeSearchData,
} from "./recipeSearchHelpers.ts";
import {
    addIngredientsToShoppingList,
    categorizeRecipesByMealType,
    createMealPlanConsoleOutput,
    createMealPlanDocument,
    fetchRecipeDetails,
    generateMealPlanStructure,
    type Recipe,
    saveMealPlanDocument,
} from "./mealPlanHelpers.ts";

const MCP_BASE_URL = Deno.env.get("MCP_BASE_URL") || "http://localhost:8001";

// Tool for searching recipes by name with guided discovery
const searchRecipesByRecipeNameTool = new DynamicStructuredTool({
    name: "search_recipes_by_recipe_name",
    description: "Implementuje workflow pro vyhledávání receptů podle názvu. " +
        "PRVNÍ KROK: Pokud uživatel chce recepty podle názvu, ale nespecifikoval konkrétní název, nastav 'show_available_names' na true pro zobrazení všech dostupných názvů receptů. " +
        "DRUHÝ KROK: Poté co uživatel vybere konkrétní název z seznamu, hledej podle tohoto názvu. " +
        "DŮLEŽITÉ!!!: NIKDY NEVYPISUJ UŽIVATELI CELÝ SEZNAM RECEPTŮ! JE PŘÍLIŠ DLOUHÝ! " +
        "Podporuje částečnou shodu a stránkování výsledků (max 10 receptů na stránku).",
    schema: z.object({
        show_available_names: z
            .boolean()
            .optional()
            .describe(
                "Nastav na true pro zobrazení všech dostupných názvů receptů jako první krok",
            ),
        name: z
            .string()
            .optional()
            .describe(
                "Konkrétní název receptu pro vyhledání (použij po zobrazení dostupných názvů)",
            ),
        page: z
            .number()
            .optional()
            .describe(
                "Číslo stránky pro stránkování výsledků (výchozí: 1, max 10 receptů na stránku)",
            ),
    }),
    func: async ({ show_available_names, name, page }) => {
        try {
            // STEP 1: Show available recipe names if requested
            if (show_available_names && !name) {
                try {
                    const namesData = await fetchAllRecipeNames(MCP_BASE_URL);
                    return JSON.stringify(
                        {
                            message:
                                `📋 K dispozici je ${namesData.count} receptů`,
                            available_recipe_names: namesData.recipe_names,
                            instruction:
                                "Vyber si konkrétní název receptu a použij parametr 'name' pro vyhledání.",
                            note:
                                "POZOR: Seznam je dlouhý - použij Ctrl+F pro rychlé hledání konkrétního názvu.",
                        },
                        null,
                        2,
                    );
                } catch (error) {
                    return `Error fetching recipe names: ${
                        error instanceof Error ? error.message : "Unknown error"
                    }`;
                }
            }

            // STEP 2: Search by specific recipe name
            if (name) {
                const params = new URLSearchParams();
                params.append("name", name);
                if (page) params.append("page", page.toString());

                const response = await fetch(
                    `${MCP_BASE_URL}/search_recipes?${params.toString()}`,
                );

                if (!response.ok) {
                    throw new Error(`HTTP error! status: ${response.status}`);
                }

                const data = await response.json() as RecipeSearchData;

                // FALLBACK: If no recipes found, use intelligent recipe matching
                if (!data.recipes || data.recipes.length === 0) {
                    return await handleRecipeSearchFallback(
                        name,
                        MCP_BASE_URL,
                        data,
                    );
                }

                // SUCCESS: Recipes found, add pagination info and return
                const dataWithPagination = addPaginationInfo(data);
                return JSON.stringify(dataWithPagination, null, 2);
            }

            // ERROR: Neither option was used correctly
            return JSON.stringify(
                {
                    error:
                        "Musíš použít buď 'show_available_names: true' pro zobrazení dostupných názvů, nebo 'name' pro vyhledání konkrétního receptu.",
                    workflow:
                        "1. Nejdříve nastav show_available_names: true\n2. Pak použij konkrétní název z seznamu v parametru name",
                },
                null,
                2,
            );
        } catch (error) {
            return `Error in recipe name workflow: ${
                error instanceof Error ? error.message : "Unknown error"
            }`;
        }
    },
});

const searchRecipesByIngredientsTool = new DynamicStructuredTool({
    name: "search_recipes_by_ingredients",
    description:
        "Hledej recepty podle ingrediencí s inteligentním fallbackem. " +
        "PRVNÍ POKUS: Nejdříve hledej přímo podle zadané ingredience. " +
        "FALLBACK: Pokud nenajde žádné recepty, zobrazí dostupné ingredience pro nalezení podobných názvů. " +
        "Použij když uživatel hledá recepty s konkrétní ingrediencí (např. 'najdi mi recept se sýrem Gruyere').",
    schema: z.object({
        ingredient: z
            .string()
            .describe(
                "Název ingredience pro vyhledání receptů (např. 'Gruyere', 'česnek', 'rajčata')",
            ),
        fallback_ingredient: z
            .string()
            .optional()
            .describe(
                "Přesný název ingredience z seznamu dostupných ingrediencí (použij po fallbacku když první pokus neuspěl)",
            ),
        page: z
            .number()
            .optional()
            .describe(
                "Číslo stránky pro stránkování výsledků (výchozí: 1, max 10 receptů na stránku)",
            ),
    }),
    func: async ({ ingredient, fallback_ingredient, page }) => {
        try {
            return await handleIngredientSearch(
                ingredient,
                fallback_ingredient,
                page,
                MCP_BASE_URL,
            );
        } catch (error) {
            return `Error in ingredient search workflow: ${
                error instanceof Error ? error.message : "Unknown error"
            }`;
        }
    },
});

// Tool for searching recipes by diet with intelligent fallback
const searchRecipesByDietTool = new DynamicStructuredTool({
    name: "search_recipes_by_diet",
    description:
        "Hledej recepty podle diety s inteligentním fallbackem a podporou stránkování. " +
        "PRVNÍ POKUS: Nejdříve hledej přímo podle zadané diety. " +
        "STRÁNKOVÁNÍ: Pokud agent nenajde dostatek receptů, může listovat dalšími stránkami. " +
        "FALLBACK: Pokud nenajde žádné recepty, zobrazí dostupné diety pro nalezení podobných názvů. " +
        "Použij když uživatel hledá recepty podle diety (např. 'najdi mi vegetariánské recepty', 'bezlepkové jídlo').",
    schema: z.object({
        diet: z
            .string()
            .describe(
                "Název diety pro vyhledání receptů (např. 'vegetarian', 'bezlepkové', 'vegan')",
            ),
        fallback_diet: z
            .string()
            .optional()
            .describe(
                "Přesný název diety z seznamu dostupných diet (použij po fallbacku když první pokus neuspěl)",
            ),
        page: z
            .number()
            .optional()
            .describe(
                "Číslo stránky pro stránkování výsledků (výchozí: 1, max 10 receptů na stránku). Použij pro listování dalšími stránkami.",
            ),
    }),
    func: async ({ diet, fallback_diet, page }) => {
        try {
            return await handleDietSearch(
                diet,
                fallback_diet,
                page,
                MCP_BASE_URL,
            );
        } catch (error) {
            return `Error in diet search workflow: ${
                error instanceof Error ? error.message : "Unknown error"
            }`;
        }
    },
});

// Tool for searching recipes by criteria (excluding name)
const searchRecipesTool = new DynamicStructuredTool({
    name: "search_recipes",
    description:
        "Hledej recepty podle diety (diet), typu jídle nebo chodu (meal_type), nebo ingrediencí. Parametry vyhledávání se dají kombinovat. " +
        "Podporuje filtrování podle ingrediencí které mají/nemají být v receptu a stránkování výsledků (max 10 receptů na stránku). " +
        "Vrací recepty s informacemi o stránkování - pokud je více výsledků, použij parametr 'page' pro načtení dalších stránek. " +
        "Pro vyhledávání podle názvu použij search_recipes_by_name.",
    schema: z.object({
        diet: z
            .string()
            .optional()
            .describe(
                "Vrátí filtrované recepty podle diety nebo kategorie stravování. Možnosti: 'bez laktozy', 'bezlepkové', 'high-protein', 'low-carb', 'masité', 'tučné', 'vegan', 'vegetarian'",
            ),
        meal_type: z
            .string()
            .optional()
            .describe(
                "Vrátí filtrované recepty podle typu jídla (meal type), Možnosti: 'desert', 'dochucovadlo', 'hlavní chod', 'polévka', 'pomazánka', 'předkrm', 'příloha', 'salát', 'snídaně'",
            ),
        includes_ingredients: z
            .string()
            .optional()
            .describe(
                "Seznam ingrediencí oddělených čárkou, které MUSÍ být v receptu (např. 'Cibule,Máslo,Česnek')",
            ),
        excludes_ingredients: z
            .string()
            .optional()
            .describe(
                "Seznam ingrediencí oddělených čárkou, které NESMÍ být v receptu (např. 'Mléko,Vejce')",
            ),
        page: z
            .number()
            .optional()
            .describe(
                "Číslo stránky pro stránkování výsledků (výchozí: 1, max 10 receptů na stránku)",
            ),
    }),
    func: async (
        {
            diet,
            meal_type,
            includes_ingredients,
            excludes_ingredients,
            page,
        },
    ) => {
        try {
            const params = new URLSearchParams();
            if (diet) params.append("diet", diet);
            if (meal_type) params.append("meal_type", meal_type);
            if (includes_ingredients) {
                params.append("includes_ingredients", includes_ingredients);
            }
            if (excludes_ingredients) {
                params.append("excludes_ingredients", excludes_ingredients);
            }
            if (page) params.append("page", page.toString());

            const response = await fetch(
                `${MCP_BASE_URL}/search_recipes?${params.toString()}`,
            );

            if (!response.ok) {
                throw new Error(`HTTP error! status: ${response.status}`);
            }

            const data = await response.json() as RecipeSearchData;

            // Handle pagination info for better user experience using helper
            const dataWithPagination = addPaginationInfo(data);
            return JSON.stringify(dataWithPagination, null, 2);
        } catch (error) {
            return `Error searching recipes: ${
                error instanceof Error ? error.message : "Unknown error"
            }`;
        }
    },
});

// Tool for adding ingredients to shopping list
const addIngredientsToShoppingListTool = new DynamicStructuredTool({
    name: "add_ingredients_to_shopping_list",
    description:
        "Přidá více ingrediencí na nákupní seznam (shoping list). Užitečné při plánování jídel nebo když si uživatelé přejí přidat konkrétní položky.",
    schema: z.object({
        ingredients: z
            .array(z.string())
            .describe("Array s názvy ingrediencí k přidání na nákupní seznam"),
    }),
    func: async ({ ingredients }) => {
        try {
            const response = await fetch(`${MCP_BASE_URL}/add_ingredients`, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify({ ingredients }),
            });

            if (!response.ok) {
                throw new Error(`HTTP error! status: ${response.status}`);
            }

            const data = await response.json();
            return JSON.stringify(data, null, 2);
        } catch (error) {
            return `Error adding ingredients: ${
                error instanceof Error ? error.message : "Unknown error"
            }`;
        }
    },
});

// Tool for getting the current shopping list
const getShoppingListTool = new DynamicStructuredTool({
    name: "get_shopping_list",
    description: "Vrátí obsah aktuálního nákupního seznamu se všemi položkami.",
    schema: z.object({}),
    func: async () => {
        try {
            const response = await fetch(`${MCP_BASE_URL}/get_shopping_list`);

            if (!response.ok) {
                throw new Error(`HTTP error! status: ${response.status}`);
            }

            const data = await response.json();
            const list = JSON.stringify(data, null, 2);
            return list;
        } catch (error) {
            return `Error getting shopping list: ${
                error instanceof Error ? error.message : "Unknown error"
            }`;
        }
    },
});

// Tool for clearing the shopping list
export const clearShoppingListTool = new DynamicStructuredTool({
    name: "clear_shopping_list",
    description:
        "Odstraní všechny položky z nákupního seznamu. Použij, když chce uživatel začít znovu nebo již dokončil nákup.",
    schema: z.object({}),
    func: async () => {
        try {
            const response = await fetch(
                `${MCP_BASE_URL}/clear_shopping_list`,
                {
                    method: "POST",
                    headers: {
                        "Content-Type": "application/json",
                    },
                },
            );

            if (!response.ok) {
                throw new Error(`HTTP error! status: ${response.status}`);
            }

            const data = await response.json();
            return JSON.stringify(data, null, 2);
        } catch (error) {
            return `Error clearing shopping list: ${
                error instanceof Error ? error.message : "Unknown error"
            }`;
        }
    },
});

// Tool for removing ingredients from shopping list
const removeIngredientsFromShoppingListTool = new DynamicStructuredTool({
    name: "remove_ingredients_from_shopping_list",
    description:
        "Odstraní specifické ingredience z nákupního seznamu. Ingredience, které nejsou v seznamu, budou ignorovány. Užitečné pro úpravu nákupního seznamu nebo když se uživatel rozhodne některé položky nechtít.",
    schema: z.object({
        ingredients: z
            .array(z.string())
            .describe(
                "Array s názvy ingrediencí k odstranění z nákupního seznamu",
            ),
    }),
    func: async ({ ingredients }) => {
        try {
            const response = await fetch(`${MCP_BASE_URL}/remove_ingredients`, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify({ ingredients }),
            });

            if (!response.ok) {
                throw new Error(`HTTP error! status: ${response.status}`);
            }

            const data = await response.json();
            return JSON.stringify(data, null, 2);
        } catch (error) {
            return `Error removing ingredients: ${
                error instanceof Error ? error.message : "Unknown error"
            }`;
        }
    },
});

// Tool for creating meal plans from diet with intelligent recipe selection
const createMealPlanFromDietTool = new DynamicStructuredTool({
    name: "create_meal_plan_from_diet",
    description: "Vytvoří kompletní jídelníček na základě diety a počtu dní. " +
        "INTELIGENTNÍ WORKFLOW: 1) Automaticky vyhledá recepty podle zadané diety " +
        "2) Inteligentně vybere vhodné recepty pro každý typ jídla (snídaně, oběd, večeře, svačina) " +
        "3) Vytvoří strukturovaný dokument s kompletním jídelníčkem včetně všech ingrediencí a postupů " +
        "4) Volitelně přidá všechny ingredience na nákupní seznam " +
        "Toto je hlavní nástroj pro vytváření jídelníčků podle požadavků z ASSIGNMENT.md!",
    schema: z.object({
        diet: z
            .string()
            .describe(
                "Název diety pro vyhledání receptů (např. 'vegetarian', 'vegan', 'bezlepkové', 'low-carb')",
            ),
        days: z
            .number()
            .min(1)
            .max(14)
            .describe("Počet dní jídelníčku (1-14 dní)"),
        meals_per_day: z
            .array(z.enum(["snídaně", "oběd", "večeře", "svačina"]))
            .describe(
                "Typy jídel pro každý den (např. ['snídaně', 'oběd', 'večeře'] nebo ['oběd', 'večeře'])",
            ),
        title: z
            .string()
            .optional()
            .describe(
                "Vlastní název jídelníčku (pokud není zadán, vytvoří se automaticky)",
            ),
        add_to_shopping_list: z
            .boolean()
            .optional()
            .describe(
                "Zda přidat všechny ingredience z jídelníčku do nákupního seznamu (výchozí: false)",
            ),
        fallback_diet: z
            .string()
            .optional()
            .describe(
                "Přesný název diety z seznamu dostupných diet (použij po fallbacku když první pokus neuspěl)",
            ),
    }),
    func: async ({
        diet,
        days,
        meals_per_day,
        title,
        add_to_shopping_list = false,
        fallback_diet,
    }) => {
        try {
            // Search for recipes using the diet search workflow
            const searchResult = await fetchRecipesByDietWithPagination(
                diet,
                fallback_diet,
                MCP_BASE_URL,
                days * meals_per_day.length * 2, // Get enough recipes for variety
            );

            // Check if we got a fallback response (no recipes found)
            if ("fallbackResponse" in searchResult) {
                return searchResult.fallbackResponse;
            }

            const { recipes: allRecipes, searchDiet } = searchResult;

            if (allRecipes.length === 0) {
                return `❌ Nepodařilo se najít žádné recepty pro dietu "${searchDiet}". Zkus použít jiný název diety nebo použij search_recipes_by_diet pro zobrazení dostupných diet.`;
            }

            // Categorize recipes by meal type for intelligent distribution
            const recipesByMealType = categorizeRecipesByMealType(
                allRecipes as Recipe[],
            );

            // Generate meal plan structure
            const mealPlanDays = generateMealPlanStructure(
                days,
                meals_per_day,
                recipesByMealType,
            );

            // Create the document title
            const mealPlanTitle = title ||
                `${
                    searchDiet.charAt(0).toUpperCase() + searchDiet.slice(1)
                } jídelníček na ${days} ${
                    days === 1 ? "den" : days < 5 ? "dny" : "dní"
                }`;

            // Collect all unique recipe names and fetch detailed information
            const allRecipeNames = new Set<string>();
            mealPlanDays.forEach((day) => {
                day.meals.forEach((meal) => {
                    allRecipeNames.add(meal.recipe_name);
                });
            });

            const recipeDetails = await fetchRecipeDetails(
                allRecipeNames,
                MCP_BASE_URL,
            );

            // Create the formatted meal plan document
            const { content, ingredientsCount, recipesCount } =
                createMealPlanDocument(
                    mealPlanTitle,
                    searchDiet,
                    days,
                    meals_per_day,
                    mealPlanDays,
                    recipeDetails,
                );

            // Save the document to file
            const filename = await saveMealPlanDocument(content);

            // Optionally add ingredients to shopping list
            let shoppingListResult = "";
            if (add_to_shopping_list) {
                shoppingListResult = await addIngredientsToShoppingList(
                    recipeDetails,
                    MCP_BASE_URL,
                );
            }

            // Create user-friendly console output
            const consoleOutput = createMealPlanConsoleOutput(
                mealPlanTitle,
                filename,
                searchDiet,
                days,
                recipesCount,
                ingredientsCount,
                mealPlanDays,
                shoppingListResult,
            );

            return consoleOutput;
        } catch (error) {
            return `❌ Chyba při vytváření jídelníčku: ${
                error instanceof Error ? error.message : "Neznámá chyba"
            }`;
        }
    },
});

// Export all MCP tools as an array
export const mcpTools = [
    searchRecipesByRecipeNameTool,
    searchRecipesByIngredientsTool,
    searchRecipesByDietTool,
    searchRecipesTool,
    addIngredientsToShoppingListTool,
    removeIngredientsFromShoppingListTool,
    getShoppingListTool,
    clearShoppingListTool,
    createMealPlanFromDietTool,
];

File: src/tools/mcp_tools.py
import os
import httpx
from typing import List, Optional, Dict, Any
from langchain_core.tools import tool
from pydantic import BaseModel, Field

# Get MCP server URL from environment
MCP_BASE_URL = os.getenv("MCP_BASE_URL", "http://localhost:8001")

class SearchRecipesInput(BaseModel):
    """Input schema for recipe search"""
    diet: Optional[str] = Field(
        None,
        description="Filter recipes by diet or food category. Options: 'bez laktozy', 'bezlepkové', 'high-protein', 'low-carb', 'masité', 'tučné', 'vegan', 'vegetarian'"
    )
    meal_type: Optional[str] = Field(
        None,
        description="Filter recipes by meal type. Options: 'desert', 'dochucovadlo', 'hlavní chod', 'polévka', 'pomazánka', 'předkrm', 'příloha', 'salát', 'snídaně'"
    )
    name: Optional[str] = Field(
        None,
        description="Search recipes by name (partial match)"
    )

class IngredientsInput(BaseModel):
    """Input schema for ingredients operations"""
    ingredients: List[str] = Field(
        description="Array of ingredient names to add/remove from shopping list"
    )

class MealPlanDay(BaseModel):
    """Schema for a single day in meal plan"""
    day_name: str = Field(description="Name of the day (e.g., 'Den 1 - Pondělí')")
    meals: List[Dict[str, str]] = Field(
        description="List of meals for the day with meal_type and recipe_name"
    )

class CreateMealPlanInput(BaseModel):
    """Input schema for meal plan creation"""
    title: str = Field(description="Title of the meal plan (e.g., 'Vegetariánský jídelníček na týden')")
    days: List[MealPlanDay] = Field(description="Array of days with their meals")

@tool
def search_recipes(diet: Optional[str] = None, meal_type: Optional[str] = None, name: Optional[str] = None) -> str:
    """
    Search for recipes by diet, meal type, or name. Parameters can be combined.
    Useful when you want to find recipes by specific criteria.
    If no recipes are found, you can use get_all_recipes endpoint.
    """
    try:
        params = {}
        if diet:
            params["diet"] = diet
        if meal_type:
            params["meal_type"] = meal_type
        if name:
            params["name"] = name
            
        with httpx.Client() as client:
            response = client.get(f"{MCP_BASE_URL}/search_recipes", params=params)
            response.raise_for_status()
            data = response.json()
            
        import json
        return json.dumps(data, indent=2, ensure_ascii=False)
    except Exception as error:
        return f"Error searching recipes: {str(error)}"

@tool
def get_all_recipes() -> str:
    """
    Returns a list of all available recipes in the database.
    The recipe list is quite long, so use this tool only if you don't find any recipes through search_recipes endpoint.
    """
    try:
        with httpx.Client() as client:
            response = client.get(f"{MCP_BASE_URL}/get_recipes")
            response.raise_for_status()
            data = response.json()
            
        import json
        return json.dumps(data, indent=2, ensure_ascii=False)
    except Exception as error:
        return f"Error getting recipes: {str(error)}"

@tool
def add_ingredients_to_shopping_list(ingredients: List[str]) -> str:
    """
    Add multiple ingredients to the shopping list. 
    Useful when planning meals or when users want to add specific items.
    """
    try:
        payload = {"ingredients": ingredients}
        
        with httpx.Client() as client:
            response = client.post(
                f"{MCP_BASE_URL}/add_ingredients",
                json=payload,
                headers={"Content-Type": "application/json"}
            )
            response.raise_for_status()
            data = response.json()
            
        import json
        return json.dumps(data, indent=2, ensure_ascii=False)
    except Exception as error:
        return f"Error adding ingredients: {str(error)}"

@tool
def get_shopping_list() -> str:
    """
    Returns the content of current shopping list with all items.
    """
    try:
        with httpx.Client() as client:
            response = client.get(f"{MCP_BASE_URL}/get_shopping_list")
            response.raise_for_status()
            data = response.json()
            
        import json
        return json.dumps(data, indent=2, ensure_ascii=False)
    except Exception as error:
        return f"Error getting shopping list: {str(error)}"

@tool
def clear_shopping_list() -> str:
    """
    Remove all items from the shopping list. 
    Use when user wants to start over or has completed shopping.
    """
    try:
        with httpx.Client() as client:
            response = client.post(
                f"{MCP_BASE_URL}/clear_shopping_list",
                headers={"Content-Type": "application/json"}
            )
            response.raise_for_status()
            data = response.json()
            
        import json
        return json.dumps(data, indent=2, ensure_ascii=False)
    except Exception as error:
        return f"Error clearing shopping list: {str(error)}"

@tool
def remove_ingredients_from_shopping_list(ingredients: List[str]) -> str:
    """
    Remove specific ingredients from the shopping list. 
    Ingredients not in the list will be ignored.
    Useful for editing shopping list or when user decides not to want some items.
    """
    try:
        payload = {"ingredients": ingredients}
        
        with httpx.Client() as client:
            response = client.post(
                f"{MCP_BASE_URL}/remove_ingredients",
                json=payload,
                headers={"Content-Type": "application/json"}
            )
            response.raise_for_status()
            data = response.json()
            
        import json
        return json.dumps(data, indent=2, ensure_ascii=False)
    except Exception as error:
        return f"Error removing ingredients: {str(error)}"

@tool
def create_meal_plan(title: str, days: List[Dict[str, Any]]) -> str:
    """
    Create a structured meal plan for multiple days and save it as a markdown file.
    Use this tool after creating a meal plan for several days ahead.
    """
    meal_emojis = {
        "snídaně": "🥐",
        "oběd": "🍽️", 
        "večeře": "🌙",
        "svačina": "🍪"
    }
    
    try:
        # Collect all unique recipe names
        all_recipe_names = set()
        for day in days:
            for meal in day.get("meals", []):
                all_recipe_names.add(meal.get("recipe_name", ""))
        
        # Fetch recipe details
        recipe_details = {}
        
        for recipe_name in all_recipe_names:
            if not recipe_name:
                continue
                
            try:
                with httpx.Client() as client:
                    response = client.get(
                        f"{MCP_BASE_URL}/search_recipes",
                        params={"name": recipe_name}
                    )
                    
                if response.status_code == 200:
                    data = response.json()
                    if data.get("recipes") and len(data["recipes"]) > 0:
                        recipe_details[recipe_name] = data["recipes"][0]
                    else:
                        print(f"LOG: No recipe found for \"{recipe_name}\" ⚠️")
                        recipe_details[recipe_name] = {
                            "name": recipe_name,
                            "ingredients": [],
                            "steps": "Recept nebyl nalezen v databázi."
                        }
            except Exception as error:
                print(f"LOG: Error fetching recipe \"{recipe_name}\": {error}")
                recipe_details[recipe_name] = {
                    "name": recipe_name,
                    "ingredients": [],
                    "steps": "Chyba při načítání receptu."
                }
        
        # Create meal plan text
        meal_plan_text = f"# {title}\n\n"
        
        for day in days:
            meal_plan_text += f"🗓️ **{day['day_name']}:**\n"
            
            # Sort meals by type
            meal_order = ["snídaně", "oběd", "večeře", "svačina"]
            sorted_meals = sorted(
                day.get("meals", []),
                key=lambda x: meal_order.index(x.get("meal_type", "")) if x.get("meal_type") in meal_order else 999
            )
            
            for meal in sorted_meals:
                meal_type = meal.get("meal_type", "")
                recipe_name = meal.get("recipe_name", "")
                emoji = meal_emojis.get(meal_type, "🍽️")
                capitalized_meal_type = meal_type.capitalize() if meal_type else ""
                meal_plan_text += f"  • {emoji} {capitalized_meal_type}: {recipe_name}\n"
                
            meal_plan_text += "\n"
        
        meal_plan_text += "---\n\n## Recepty\n\n"
        
        # Add detailed recipes
        found_recipes = {
            name: recipe for name, recipe in recipe_details.items()
            if recipe.get("ingredients") and len(recipe["ingredients"]) > 0
        }
        
        for recipe in found_recipes.values():
            meal_plan_text += f"### {recipe['name']}\n\n"
            
            if recipe.get("ingredients"):
                meal_plan_text += "**Ingredience:**\n"
                for ingredient in recipe["ingredients"]:
                    meal_plan_text += f"- {ingredient}\n"
                meal_plan_text += "\n"
            
            if recipe.get("steps"):
                meal_plan_text += f"**Postup:**\n{recipe['steps']}\n\n"
        
        # Add timestamp
        from datetime import datetime
        timestamp = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
        meal_plan_text += f"*Jídelníček vytvořen: {datetime.now().strftime('%d.%m.%Y %H:%M')}*\n"
        
        # Create plans directory and save file
        import os
        plans_dir = "./plans"
        os.makedirs(plans_dir, exist_ok=True)
        
        filename = f"jidelnicek_{timestamp}.md"
        filepath = os.path.join(plans_dir, filename)
        
        with open(filepath, "w", encoding="utf-8") as f:
            f.write(meal_plan_text)
            
        print(f"💾 Kompletní jídelníček s {len(all_recipe_names)} recepty byl uložen jako: plans/{filename}")
        
        # Create console output
        console_output = f"📅 JÍDELNÍČEK: {title}\n\n"
        
        for day in days:
            console_output += f"🗓️ {day['day_name']}:\n"
            
            # Group meals by type
            meals_by_type = {}
            for meal in day.get("meals", []):
                meal_type = meal.get("meal_type", "")
                recipe_name = meal.get("recipe_name", "")
                if meal_type not in meals_by_type:
                    meals_by_type[meal_type] = []
                meals_by_type[meal_type].append(recipe_name)
            
            # Display in preferred order
            meal_order = ["snídaně", "oběd", "večeře", "svačina"]
            for meal_type in meal_order:
                if meal_type in meals_by_type:
                    capitalized_type = meal_type.capitalize()
                    recipes = ", ".join(meals_by_type[meal_type])
                    console_output += f"  • {capitalized_type}: {recipes}\n"
            
            console_output += "\n"
        
        return console_output
        
    except Exception as error:
        return f"Error creating meal plan: {str(error)}"

# List of all MCP tools
mcp_tools = [
    search_recipes,
    get_all_recipes,
    add_ingredients_to_shopping_list,
    remove_ingredients_from_shopping_list,
    get_shopping_list,
    clear_shopping_list,
    create_meal_plan
]
